File: scaffold/src/main/java/top/xuqingquan/web/AgentWebUtils.java
package top.xuqingquan.web;

import android.app.Activity;
import android.content.Context;
import android.content.Intent;
import android.content.pm.ApplicationInfo;
import android.content.pm.PackageManager;
import android.net.Uri;
import android.os.Build;
import android.os.Environment;
import android.os.Handler;
import android.os.Looper;
import android.provider.MediaStore;
import android.text.SpannableString;
import android.text.Spanned;
import android.text.TextUtils;
import android.text.style.ForegroundColorSpan;
import android.view.View;
import android.view.ViewGroup;
import android.view.ViewParent;
import android.webkit.ValueCallback;
import android.webkit.WebChromeClient;
import android.webkit.WebSettings;
import android.webkit.WebView;
import android.widget.Toast;
import androidx.annotation.ColorInt;
import androidx.annotation.NonNull;
import androidx.annotation.RequiresApi;
import androidx.core.app.AppOpsManagerCompat;
import androidx.core.content.ContextCompat;
import androidx.core.os.EnvironmentCompat;
import com.google.android.material.snackbar.Snackbar;
import top.xuqingquan.R;
import top.xuqingquan.utils.FileUtils;
import top.xuqingquan.utils.Timber;

import java.io.File;
import java.io.IOException;
import java.lang.ref.WeakReference;
import java.lang.reflect.Method;
import java.text.SimpleDateFormat;
import java.util.*;

import static top.xuqingquan.web.AgentWebConfig.AGENTWEB_FILE_PATH;
import static top.xuqingquan.web.AgentWebConfig.FILE_CACHE_PATH;

@SuppressWarnings("ResultOfMethodCallIgnored")
public class AgentWebUtils {

    private static Handler mHandler = null;

    private AgentWebUtils() {
        throw new UnsupportedOperationException("u can't init me");
    }

    public static int dp2px(Context context, float dipValue) {
        final float scale = context.getResources().getDisplayMetrics().density;
        return (int) (dipValue * scale + 0.5f);
    }

    static void clearWebView(WebView m) {
        if (m == null) {
            return;
        }
        if (!isUIThread()) {
            return;
        }
        m.loadUrl("about:blank");
        m.stopLoading();
        if (m.getHandler() != null) {
            m.getHandler().removeCallbacksAndMessages(null);
        }
        m.removeAllViews();
        ViewGroup mViewGroup = ((ViewGroup) m.getParent());
        if (mViewGroup != null) {
            mViewGroup.removeView(m);
        }
        m.setWebChromeClient(null);
        m.setWebViewClient(null);
        m.setTag(null);
        m.clearHistory();
        m.destroy();
    }

    public static String getAgentWebFilePath(Context context) {
        if (!TextUtils.isEmpty(AGENTWEB_FILE_PATH)) {
            return AGENTWEB_FILE_PATH;
        }
        String dir = getDiskExternalCacheDir(context);
        File mFile = new File(dir, FILE_CACHE_PATH);
        try {
            if (!mFile.exists()) {
                mFile.mkdirs();
            }
        } catch (Throwable throwable) {
            Timber.i("create dir exception");
        }
        Timber.i("path:" + mFile.getAbsolutePath() + "  path:" + mFile.getPath());
        AGENTWEB_FILE_PATH = mFile.getAbsolutePath();
        return AGENTWEB_FILE_PATH;
    }


    public static File createFileByName(Context context, String name, boolean cover) throws IOException {
        String path = getAgentWebFilePath(context);
        if (TextUtils.isEmpty(path)) {
            return null;
        }
        File mFile = new File(path, name);
        if (mFile.exists()) {
            if (cover) {
                mFile.delete();
                mFile.createNewFile();
            }
        } else {
            mFile.createNewFile();
        }
        return mFile;
    }

    static String getDiskExternalCacheDir(Context context) {
        File mFile = context.getExternalCacheDir();
        if (mFile != null && Environment.MEDIA_MOUNTED.equals(EnvironmentCompat.getStorageState(mFile))) {
            return mFile.getAbsolutePath();
        }
        return null;
    }

    private static WeakReference<Snackbar> snackbarWeakReference;

    static void show(View parent,
                     CharSequence text,
                     int duration,
                     @ColorInt int textColor,
                     @ColorInt int bgColor,
                     CharSequence actionText,
                     @ColorInt int actionTextColor,
                     View.OnClickListener listener) {
        SpannableString spannableString = new SpannableString(text);
        ForegroundColorSpan colorSpan = new ForegroundColorSpan(textColor);
        spannableString.setSpan(colorSpan, 0, spannableString.length(), Spanned.SPAN_EXCLUSIVE_EXCLUSIVE);
        snackbarWeakReference = new WeakReference<>(Snackbar.make(parent, spannableString, duration));
        Snackbar snackbar = snackbarWeakReference.get();
        View view = snackbar.getView();
        view.setBackgroundColor(bgColor);
        if (actionText != null && actionText.length() > 0 && listener != null) {
            snackbar.setActionTextColor(actionTextColor);
            snackbar.setAction(actionText, listener);
        }
        snackbar.show();
    }

    static Method isExistMethod(Object o, String methodName, Class... clazzs) {
        if (null == o) {
            return null;
        }
        try {
            Class clazz = o.getClass();
            Method mMethod = clazz.getDeclaredMethod(methodName, clazzs);
            mMethod.setAccessible(true);
            return mMethod;
        } catch (Throwable throwable) {
            Timber.e(throwable);
        }
        return null;

    }

    static void clearWebViewAllCache(Context context, WebView webView) {
        try {
            AgentWebConfig.removeAllCookies(null);
            webView.getSettings().setCacheMode(WebSettings.LOAD_NO_CACHE);
            context.deleteDatabase("webviewCache.db");
            context.deleteDatabase("webview.db");
            webView.clearCache(true);
            webView.clearHistory();
            webView.clearFormData();
            FileUtils.clearCacheFolder(new File(AgentWebConfig.getCachePath(context)), 0);
        } catch (Throwable t) {
            Timber.e(t);
        }
    }

    static void clearWebViewAllCache(Context context) {
        try {
            clearWebViewAllCache(context, new WebView(context.getApplicationContext()));
        } catch (Exception e) {
            e.printStackTrace();
        }
    }

    static File createImageFile(Context context) {
        File mFile = null;
        try {
            String timeStamp =
                    new SimpleDateFormat("yyyyMMddHHmmss", Locale.getDefault()).format(new Date());
            String imageName = String.format("aw_%s.jpg", timeStamp);
            mFile = createFileByName(context, imageName, true);
        } catch (Throwable e) {
            e.printStackTrace();
        }
        return mFile;
    }

    static Intent getIntentCaptureCompat(Context context, File file) {
        Intent mIntent = new Intent(MediaStore.ACTION_IMAGE_CAPTURE);
        Uri mUri = FileUtils.getUriFromFile(context, file);
        mIntent.addCategory(Intent.CATEGORY_DEFAULT);
        mIntent.putExtra(MediaStore.EXTRA_OUTPUT, mUri);
        return mIntent;
    }

    public static boolean isUIThread() {
        return Looper.myLooper() == Looper.getMainLooper();
    }

    private static Toast mToast = null;

    static void toastShowShort(Context context, String msg) {
        if (mToast == null) {
            mToast = Toast.makeText(context.getApplicationContext(), msg, Toast.LENGTH_SHORT);
        } else {
            mToast.setText(msg);
        }
        mToast.show();
    }

    @Deprecated
    static void getUIControllerAndShowMessage(Activity activity, String message, String from) {
        if (activity == null || activity.isFinishing()) {
            return;
        }
        WebParentLayout mWebParentLayout = activity.findViewById(R.id.web_parent_layout_id);
        AbsAgentWebUIController mAgentWebUIController = mWebParentLayout.provide();
        if (mAgentWebUIController != null) {
            mAgentWebUIController.onShowMessage(message, from);
        }
    }

    public static boolean hasPermission(@NonNull Context context, @NonNull String... permissions) {
        return hasPermission(context, Arrays.asList(permissions));
    }

    public static boolean hasPermission(@NonNull Context context, @NonNull List<String> permissions) {
        if (Build.VERSION.SDK_INT < Build.VERSION_CODES.M) {
            return true;
        }
        for (String permission : permissions) {
            int result = ContextCompat.checkSelfPermission(context, permission);
            if (result == PackageManager.PERMISSION_DENIED) {
                return false;
            }
            String op = AppOpsManagerCompat.permissionToOp(permission);
            if (TextUtils.isEmpty(op)) {
                continue;
            }
            result = AppOpsManagerCompat.noteProxyOp(context, op, context.getPackageName());
            if (result != AppOpsManagerCompat.MODE_ALLOWED) {
                return false;
            }
        }
        return true;
    }

    public static List<String> getDeniedPermissions(Activity activity, String[] permissions) {
        if (permissions == null || permissions.length == 0) {
            return null;
        }
        List<String> deniedPermissions = new ArrayList<>();
        for (int i = 0; i < permissions.length; i++) {
            if (!hasPermission(activity, permissions[i])) {
                deniedPermissions.add(permissions[i]);
            }
        }
        return deniedPermissions;
    }


    public static AbsAgentWebUIController getAgentWebUIControllerByWebView(WebView webView) {
        WebParentLayout mWebParentLayout = getWebParentLayoutByWebView(webView);
        return mWebParentLayout.provide();
    }

    //获取应用的名称
    public static String getApplicationName(Context context) {
        PackageManager packageManager = null;
        ApplicationInfo applicationInfo;
        try {
            packageManager = context.getApplicationContext().getPackageManager();
            applicationInfo = packageManager.getApplicationInfo(context.getPackageName(), 0);
        } catch (PackageManager.NameNotFoundException e) {
            applicationInfo = null;
        }
        return packageManager.getApplicationLabel(applicationInfo).toString();
    }

    static WebParentLayout getWebParentLayoutByWebView(WebView webView) {
        ViewGroup mViewGroup;
        if (!(webView.getParent() instanceof ViewGroup)) {
            throw new IllegalStateException("please check webcreator's create method was be called ?");
        }
        mViewGroup = (ViewGroup) webView.getParent();
        while (mViewGroup != null) {

            Timber.i("ViewGroup:" + mViewGroup);
            if (mViewGroup.getId() == R.id.web_parent_layout_id) {
                WebParentLayout mWebParentLayout = (WebParentLayout) mViewGroup;
                Timber.i("found WebParentLayout");
                return mWebParentLayout;
            } else {
                ViewParent mViewParent = mViewGroup.getParent();
                if (mViewParent instanceof ViewGroup) {
                    mViewGroup = (ViewGroup) mViewParent;
                } else {
                    mViewGroup = null;
                }
            }
        }
        throw new IllegalStateException("please check webcreator's create method was be called ?");
    }

    public static void runInUiThread(Runnable runnable) {
        if (mHandler == null) {
            mHandler = new Handler(Looper.getMainLooper());
        }
        mHandler.post(runnable);
    }

    @RequiresApi(api = Build.VERSION_CODES.LOLLIPOP)
    static boolean showFileChooserCompat(Activity activity,
                                         WebView webView,
                                         ValueCallback<Uri[]> valueCallbacks,
                                         WebChromeClient.FileChooserParams fileChooserParams,
                                         PermissionInterceptor permissionInterceptor,
                                         ValueCallback<Uri> valueCallback,
                                         String mimeType,
                                         Handler.Callback jsChannelCallback
    ) {
        try {
            FileChooser.Builder builder = FileChooser.newBuilder(activity, webView);
            if (valueCallbacks != null) {
                builder.setUriValueCallbacks(valueCallbacks);
            }
            if (fileChooserParams != null) {
                builder.setFileChooserParams(fileChooserParams);
            }
            if (valueCallback != null) {
                builder.setUriValueCallback(valueCallback);
            }
            if (!TextUtils.isEmpty(mimeType)) {
                builder.setAcceptType(mimeType);
            }
            if (jsChannelCallback != null) {
                builder.setJsChannelCallback(jsChannelCallback);
            }
            builder.setPermissionInterceptor(permissionInterceptor);
            builder.build().openFileChooser();
        } catch (Throwable throwable) {
            Timber.e(throwable);
            if (valueCallbacks != null) {
                Timber.i("onReceiveValue empty");
                return false;
            }
            if (valueCallback != null) {
                valueCallback.onReceiveValue(null);
            }
        }
        return true;
    }
}

File: scaffold_web/src/main/java/top/xuqingquan/web/publics/AgentWebUtils.kt
package top.xuqingquan.web.publics

import android.app.Activity
import android.content.Context
import android.net.Uri
import android.os.Build
import android.os.Handler
import android.support.annotation.ColorInt
import android.support.annotation.RequiresApi
import android.support.design.widget.Snackbar
import android.text.SpannableString
import android.text.Spanned
import android.text.TextUtils
import android.text.style.ForegroundColorSpan
import android.view.View
import android.view.ViewGroup
import android.webkit.ValueCallback
import android.webkit.WebChromeClient
import android.webkit.WebSettings
import android.webkit.WebView
import top.xuqingquan.utils.Timber
import top.xuqingquan.utils.clearCacheFolder
import top.xuqingquan.web.R
import top.xuqingquan.web.nokernel.PermissionInterceptor
import top.xuqingquan.web.nokernel.WebConfig
import top.xuqingquan.web.nokernel.WebUtils
import top.xuqingquan.web.nokernel.WebUtils.isUIThread
import java.io.File
import java.lang.ref.WeakReference
import com.tencent.smtt.sdk.ValueCallback as X5ValueCallback
import com.tencent.smtt.sdk.WebChromeClient as X5WebChromeClient
import com.tencent.smtt.sdk.WebView as X5WebView

object AgentWebUtils {

    @JvmStatic
    fun clearWebView(m: WebView?) {
        try {
            if (m == null) {
                return
            }
            if (!isUIThread()) {
                return
            }
            m.loadUrl("about:blank")
            m.stopLoading()
            if (m.handler != null) {
                m.handler.removeCallbacksAndMessages(null)
            }
            m.removeAllViews()
            val mViewGroup = m.parent as ViewGroup?
            mViewGroup?.removeView(m)
            m.webChromeClient = null
            m.webViewClient = null
            m.tag = null
            m.clearHistory()
            m.destroy()
        } catch (t: Throwable) {
        }
    }

    @JvmStatic
    fun clearWebView(m: X5WebView?) {
        try {
            if (m == null) {
                return
            }
            if (!isUIThread()) {
                return
            }
            m.loadUrl("about:blank")
            m.stopLoading()
            if (m.handler != null) {
                m.handler.removeCallbacksAndMessages(null)
            }
            m.removeAllViews()
            val mViewGroup = m.parent as ViewGroup?
            mViewGroup?.removeView(m)
            m.webChromeClient = null
            m.webViewClient = null
            m.tag = null
            m.clearHistory()
            m.destroy()
        } catch (t: Throwable) {
        }
    }

    @JvmStatic
    fun show(
        parent: View,
        text: CharSequence,
        duration: Int,
        @ColorInt textColor: Int,
        @ColorInt bgColor: Int,
        actionText: CharSequence?,
        @ColorInt actionTextColor: Int,
        listener: View.OnClickListener?
    ) {
        val spannableString = SpannableString(text)
        val colorSpan = ForegroundColorSpan(textColor)
        spannableString.setSpan(
            colorSpan,
            0,
            spannableString.length,
            Spanned.SPAN_EXCLUSIVE_EXCLUSIVE
        )
        val snackbarWeakReference = WeakReference(Snackbar.make(parent, spannableString, duration))
        val snackbar = snackbarWeakReference.get() ?: return
        snackbar.view.setBackgroundColor(bgColor)
        if (!actionText.isNullOrEmpty() && listener != null) {
            snackbar.setActionTextColor(actionTextColor)
            snackbar.setAction(actionText, listener)
        }
        snackbar.show()
    }

    @JvmStatic
    fun clearWebViewAllCache(context: Context, webView: WebView) {
        try {
            AgentWebConfig.removeAllCookies(null)
            webView.settings.cacheMode = WebSettings.LOAD_NO_CACHE
            context.deleteDatabase("webviewCache.db")
            context.deleteDatabase("webview.db")
            webView.clearCache(true)
            webView.clearHistory()
            webView.clearFormData()
            clearCacheFolder(File(WebUtils.getCachePath(context)), 0)
        } catch (t: Throwable) {
            Timber.e(t)
        }

    }

    @JvmStatic
    fun clearWebViewAllCache(context: Context, webView: X5WebView) {
        try {
            AgentWebConfig.removeAllX5Cookies(null)
            webView.settings.cacheMode = WebSettings.LOAD_NO_CACHE
            context.deleteDatabase("webviewCache.db")
            context.deleteDatabase("webview.db")
            webView.clearCache(true)
            webView.clearHistory()
            webView.clearFormData()
            clearCacheFolder(File(WebUtils.getCachePath(context)), 0)
        } catch (t: Throwable) {
            Timber.e(t)
        }

    }

    @JvmStatic
    fun clearWebViewAllCache(context: Context) {
        try {
            if (WebConfig.enableTbs()) {
                clearWebViewAllCache(
                    context,
                    top.xuqingquan.web.x5.LollipopFixedWebView(context.applicationContext)
                )
            } else {
                clearWebViewAllCache(
                    context,
                    top.xuqingquan.web.system.LollipopFixedWebView(context.applicationContext)
                )
            }
        } catch (e: Throwable) {
            e.printStackTrace()
        }

    }

    @JvmStatic
    fun getAgentWebUIControllerByWebView(webView: WebView): AbsAgentWebUIController? {
        val mWebParentLayout = getWebParentLayoutByWebView(webView)
        return mWebParentLayout.provide()
    }

    @JvmStatic
    fun getAgentWebUIControllerByWebView(webView: X5WebView): AbsAgentWebUIController? {
        val mWebParentLayout = getWebParentLayoutByWebView(webView)
        return mWebParentLayout.provide()
    }

    @JvmStatic
    fun getWebParentLayoutByWebView(webView: WebView): WebParentLayout {
        var mViewGroup: ViewGroup?
        check(webView.parent is ViewGroup) { "please check webcreator's create method was be called ?" }
        mViewGroup = webView.parent as ViewGroup
        while (mViewGroup != null) {
            Timber.i("ViewGroup:$mViewGroup")
            mViewGroup = if (mViewGroup.id == R.id.scaffold_web_parent_layout_id) {
                val mWebParentLayout = mViewGroup as WebParentLayout?
                Timber.i("found WebParentLayout")
                return mWebParentLayout!!
            } else {
                val mViewParent = mViewGroup.parent
                if (mViewParent is ViewGroup) {
                    mViewParent
                } else {
                    null
                }
            }
        }
        throw IllegalStateException("please check webcreator's create method was be called ?")
    }

    @JvmStatic
    fun getWebParentLayoutByWebView(webView: X5WebView): WebParentLayout {
        var mViewGroup: ViewGroup?
        check(webView.parent is ViewGroup) { "please check webcreator's create method was be called ?" }
        mViewGroup = webView.parent as ViewGroup
        while (mViewGroup != null) {
            Timber.i("ViewGroup:$mViewGroup")
            mViewGroup = if (mViewGroup.id == R.id.scaffold_web_parent_layout_id) {
                val mWebParentLayout = mViewGroup as WebParentLayout?
                Timber.i("found WebParentLayout")
                return mWebParentLayout!!
            } else {
                val mViewParent = mViewGroup.parent
                if (mViewParent is ViewGroup) {
                    mViewParent
                } else {
                    null
                }
            }
        }
        throw IllegalStateException("please check webcreator's create method was be called ?")
    }

    @JvmStatic
    @RequiresApi(api = Build.VERSION_CODES.LOLLIPOP)
    fun showFileChooserCompat(
        activity: Activity,
        webView: WebView,
        valueCallbacks: ValueCallback<Array<Uri>>?,
        fileChooserParams: WebChromeClient.FileChooserParams?,
        permissionInterceptor: PermissionInterceptor,
        valueCallback: ValueCallback<Uri>?,
        mimeType: String?,
        jsChannelCallback: Handler.Callback?
    ): Boolean {
        try {
            val builder = FileChooser.newBuilder(activity, webView)
            if (valueCallbacks != null) {
                builder.setUriValueCallbacks(valueCallbacks)
            }
            if (fileChooserParams != null) {
                builder.setFileChooserParams(fileChooserParams)
            }
            if (valueCallback != null) {
                builder.setUriValueCallback(valueCallback)
            }
            if (!TextUtils.isEmpty(mimeType)) {
                builder.setAcceptType(mimeType)
            }
            if (jsChannelCallback != null) {
                builder.setJsChannelCallback(jsChannelCallback)
            }
            builder.setPermissionInterceptor(permissionInterceptor)
            builder.build().openFileChooser()
        } catch (throwable: Throwable) {
            Timber.e(throwable)
            if (valueCallbacks != null) {
                Timber.i("onReceiveValue empty")
                return false
            }
            valueCallback?.onReceiveValue(null)
        }

        return true
    }

    @JvmStatic
    @RequiresApi(api = Build.VERSION_CODES.LOLLIPOP)
    fun showFileChooserCompat(
        activity: Activity,
        webView: X5WebView,
        valueCallbacks: X5ValueCallback<Array<Uri>>?,
        fileChooserParams: X5WebChromeClient.FileChooserParams?,
        permissionInterceptor: PermissionInterceptor,
        valueCallback: X5ValueCallback<Uri>?,
        mimeType: String?,
        jsChannelCallback: Handler.Callback?
    ): Boolean {
        try {
            val builder = FileChooser.newBuilder(activity, webView)
            if (valueCallbacks != null) {
                builder.setUriValueCallbacks(valueCallbacks)
            }
            if (fileChooserParams != null) {
                builder.setFileChooserParams(fileChooserParams)
            }
            if (valueCallback != null) {
                builder.setUriValueCallback(valueCallback)
            }
            if (!TextUtils.isEmpty(mimeType)) {
                builder.setAcceptType(mimeType)
            }
            if (jsChannelCallback != null) {
                builder.setJsChannelCallback(jsChannelCallback)
            }
            builder.setPermissionInterceptor(permissionInterceptor)
            builder.build().openFileChooser()
        } catch (throwable: Throwable) {
            Timber.e(throwable)
            if (valueCallbacks != null) {
                Timber.i("onReceiveValue empty")
                return false
            }
            valueCallback?.onReceiveValue(null)
        }

        return true
    }
}
